### PR TITLE
Change schedulers to chainable form

### PR DESCRIPTION
**Pull request by vincentqb**
Enable chainable schedulers as requested in #13022 by implementing the following changes:

* Changing the behavior of schedulers to the chainable formula when available (e.g. `CosineAnnealingWarmRestart` doesn't have a chainable form).
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning.
* Making `get_last_lr` the supported way of obtaining the last computed learning rate by the scheduler.
* Returning a warning referring to `get_last_lr` when invoking the `get_lr` function outside of `step`

Note that `get_lr` was used for two purposes prior to chainable schedulers: (Purpose 1) compute the new learning rate value (formal goal), (Purpose 2) read the current learning rate value (informal goal). For chainable schedulers, the two  purposes will disagree, and so two distinct functions are needed. 

Here is an example showing how to chain learning rate schedulers, and printing the learning rate at each iteration:
```python
import torch
from torch.nn import Parameter
from torch.optim import SGD
from torch.optim.lr_scheduler import ExponentialLR, StepLR

model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)

scheduler1 = ExponentialLR(optimizer, gamma=0.9)
scheduler2 = StepLR(optimizer, step_size=5, gamma=0.1)

for epoch in range(10):

    print(epoch, scheduler2.get_last_lr()[0])

    optimizer.step()
    scheduler1.step()
    scheduler2.step()
```

```
0 0.1
1 0.09000000000000001
2 0.08100000000000002
3 0.07290000000000002
4 0.06561000000000002
5 0.005904900000000002
6 0.005314410000000002
7 0.004782969000000002
8 0.004304672100000002
9 0.003874204890000002
```

# BC Breaking: Printing learning rate

```python
# Setup for codes below

import warnings
warnings.simplefilter('once', DeprecationWarning) 

import torch
from torch.nn import Parameter
from torch.optim import SGD
from torch.optim.lr_scheduler import StepLR

model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)
scheduler = StepLR(optimizer, 2)
```

As discussed in #22107, the user wants to print the learning rate at each iteration.

Version 1.3.1:
```python
for epoch in range(10):
    print(epoch, scheduler.get_lr()[0])
    optimizer.step()
    scheduler.step()
```

```
0 0.1
1 0.1
2 0.010000000000000002
3 0.010000000000000002
4 0.0010000000000000002
5 0.0010000000000000002
6 0.00010000000000000003
7 0.00010000000000000003
8 1.0000000000000003e-05
9 1.0000000000000003e-05
```

Version 1.4.0:
```python
for epoch in range(10):
    print(epoch, scheduler.get_lr()[0])
    optimizer.step()
    scheduler.step()
```

```
torch/optim/lr_scheduler.py:342: DeprecationWarning: To get the last learning rate computed by the scheduler, please use `get_last_lr()`.
  "please use `get_last_lr()`.", DeprecationWarning)
0 0.1
1 0.1
2 0.0010000000000000002
3 0.010000000000000002
4 0.00010000000000000003
5 0.0010000000000000002
6 1.0000000000000004e-05
7 0.00010000000000000003
8 1.0000000000000004e-06
9 1.0000000000000004e-05
```

Version 1.4.0:
```python
for epoch in range(10):
    print(epoch, scheduler.get_last_lr()[0])
    optimizer.step()
    scheduler.step()
```
```
0 0.1
1 0.1
2 0.010000000000000002
3 0.010000000000000002
4 0.0010000000000000002
5 0.0010000000000000002
6 0.00010000000000000003
7 0.00010000000000000003
8 1.0000000000000004e-05
9 1.0000000000000004e-05
```

When more than one schedulers are chained, a user may want to know the most recent learning rate computed by a specific scheduler. This is what `scheduler.get_last_lr()` does. Note that `optimizer.param_groups[0]['lr']` was in version 1.3.1 and remains in 1.4.0 a way of getting the current learning rate used in the optimizer.

# Deprecation: Epoch parameter

```python
# Setup for codes below

import warnings
warnings.simplefilter('once', DeprecationWarning) 

import torch
from torch.nn import Parameter
from torch.optim import SGD
from torch.optim.lr_scheduler import StepLR

model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)
scheduler = StepLR(optimizer, 2)
```

For schedulers that can be chained or that don't require `epoch`, the `epoch` parameter from `scheduler.step(epoch)` is being deprecated in favor of `scheduler.step()`. 

Version 1.3.1:
```python
for epoch in range(10):
    optimizer.step()
    scheduler.step(epoch)
    print(epoch, optimizer.param_groups[0]['lr'])
```
```
0 0.1
1 0.1
2 0.010000000000000002
3 0.010000000000000002
4 0.0010000000000000002
5 0.0010000000000000002
6 0.00010000000000000003
7 0.00010000000000000003
8 1.0000000000000003e-05
9 1.0000000000000003e-05
```

Version 1.4.0:
```python
for epoch in range(10):
    optimizer.step()
    scheduler.step(epoch)
    print(epoch, optimizer.param_groups[0]['lr'])
```

```
torch/optim/lr_scheduler.py:143: DeprecationWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, DeprecationWarning)
0 0.1
1 0.1
2 0.010000000000000002
3 0.010000000000000002
4 0.0010000000000000002
5 0.0010000000000000002
6 0.00010000000000000003
7 0.00010000000000000003
8 1.0000000000000003e-05
9 1.0000000000000003e-05
```

Version 1.4.0:
```python
for epoch in range(10):
    print(epoch, optimizer.param_groups[0]['lr'])
    optimizer.step()
    scheduler.step()
```

```
0 0.1
1 0.1
2 0.010000000000000002
3 0.010000000000000002
4 0.0010000000000000002
5 0.0010000000000000002
6 0.00010000000000000003
7 0.00010000000000000003
8 1.0000000000000003e-05
9 1.0000000000000003e-05
```

This last code would also work as expected in version 1.3.1.

If the epoch parameter was used to control the scheduler's flow, this logic will have to be moved to the training loop. For instance, as discussed in #20527:

Version 1.3.1:
```python
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
    optimizer.step()
    scheduler.step(epoch)
    print(epoch, optimizer.param_groups[0]['lr'])
```
```
0 0.1
0 0.1
1 0.1
1 0.1
2 0.010000000000000002
2 0.010000000000000002
3 0.010000000000000002
3 0.010000000000000002
```

Version 1.4.0:
```python
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
    optimizer.step()
    scheduler.step(epoch)
    print(epoch, optimizer.param_groups[0]['lr'])
```
```
torch/optim/lr_scheduler.py:143: DeprecationWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, DeprecationWarning)
0 0.1
0 0.1
1 0.1
1 0.1
2 0.010000000000000002
2 0.010000000000000002
3 0.010000000000000002
3 0.010000000000000002
```

Version 1.4.0:
```python
for epoch in range(8):
    print(epoch, optimizer.param_groups[0]['lr'])

    optimizer.step()

    # Step at every other epoch
    if epoch % 2:
        scheduler.step()
```
```
0 0.1
0 0.1
1 0.1
1 0.1
2 0.010000000000000002
2 0.010000000000000002
3 0.010000000000000002
3 0.010000000000000002
```

# Changes from #24352

This new PR is based on #24352 which was reverted. The problem was the introduction of a BC breaking change by changing the name of `get_lr`. This failed on users who built their own schedulers by modifying `get_lr`. The difference with #24352:

* In this PR, (Purpose 1) remains `get_lr`, and (Purpose 2) is now `get_last_lr`
* In #24352, (Purpose 1) was `_compute values`, and (Purpose 2) `get_last_computed_values`.

# ghstack

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27254 Add multiplicative lr.
* **#26423 Change schedulers to chainable form**

This reverts commit 1c477b7e1f378e9c1f8efed296241f68a8a4372b.

Differential Revision: [D17460427](https://our.internmc.facebook.com/intern/diff/D17460427)

### CODE REVIEW

**Comment by vadimkantorov** [comment]
```python
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)
```
this looks like magic. it seems that schedulers are attached to the optimizer, but this is not explicit. is this what's happening?

**Comment by vincentqb** [comment]
> ```python
> optimizer = torch.optim.SGD(net.parameters(), 0.1)
> lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
> lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)
> ```
> 
> this looks like magic. it seems that schedulers are attached to the optimizer, but this is not explicit. is this what's happening?

`optimizer` is the first parameter passed to `MultiStepLR` and `StepLR`, is this what you refer to?

**Comment by vadimkantorov** [comment]
I think I misunderstand that example, sorry for the bother... I thought that somehow the two created schedulers interact.

**Comment by vincentqb** [comment]
This PR (unlike #24352) preserves the name of `get_lr` but throws a warning if the users call it outside of `step`, and refers to the new interface to get the most recent learning rate, `get_last_lr`.

**Comment by ezyang** [comment]
How did you test that this no longer breaks the projects that were affected last revert? On V3 of the diff internally, it does not look you triggered deferred tests.

File: torch/optim/lr_scheduler.py
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: Wouldn't a more normal way to implement this to just have `get_latest_lr` throw no warning (and `get_lr` point at it with a warning), so you don't have to muck around with this attribute?

Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: If I understand correctly, you suggesting to just make `get_lr` throw an error when called? `get_lr` is always called within `step`, and we don't want an error thrown in this case. This attribute is meant to check that `get_lr` is called within `step` and avoid throwing an error in this case. Is that what you meant?

**Comment by vincentqb** [comment]
> How did you test that this no longer breaks the projects that were affected last revert? On V3 of the diff internally, it does not look you triggered deferred tests.

The evidence that projects no longer break is that all internal tests are passing. See T53988247 (internal) for internal tests that were failing on #24352.

The deferred internal tests are now running.

**Comment by vincentqb** [comment]
> How did you test that this no longer breaks the projects that were affected last revert? On V3 of the diff internally, it does not look you triggered deferred tests.

@ezyang -- I ran the test listed in T53988247, and passed :D

File: torch/optim/lr_scheduler.py
Comment (ezyang) [reply torch/optim/lr_scheduler.py]: > If I understand correctly, you are suggesting to just make get_lr throw an error when called?

If by "throw an error" you mean "raise a warning", yes, I am suggesting that. However, I am *also* suggesting that you change the line: 

```
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
```

To be:

```
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr_nowarn()):
```

where `get_lr_nowarn` is a version of `get_lr` that has the same semantics, but doesn't do a warning.

File: test/test_optim.py
Comment (ezyang) [inline test/test_optim.py]: Is the warning filtering here new? Why are these tests raising deprecation warnings? Shouldn't our test code be using the non-deprecated codepath?

File: test/test_optim.py
Comment (fmassa) [inline test/test_optim.py]: I wonder if it would make sense to create a `CombineScheduler` class in a follow-up PR, which takes an iterable of schedulers and apply them in order.

Basic gist

```python
class CombinedScheduler(object):
    def __init__(self, schedulers):
        # assert all schedulers are for the same optimizer?
        self.schedulers = list(schedulers)

    def step():
        for scheduler in self.schedulers:
            scheduler.step()
```

Instead of a list of schedulers, we could even pass a map of schedulers, so that at `step` call we could dynamically decide on which schedulers we want to step on, by name.

But this is out of the scope of this PR, we should probably open an issue after this PR gets merged to discuss about it.

File: torch/optim/lr_scheduler.py
Comment (fmassa) [inline torch/optim/lr_scheduler.py]: This could live in a separate PR, right?

File: torch/optim/lr_scheduler.py
Comment (fmassa) [inline torch/optim/lr_scheduler.py]: I'm ok with having it being a method called `get_last_lr`, but I wonder if it was a `@property` called `lr` it wouldn't be more natural to users? Like
```python
print(i, scheduler.lr)
```

**Review by fmassa — APPROVED** [review]
This looks good to me, thanks a lot for all the work @vincentqb !

I have made a few minor comments, none of which IMO is blocking this PR to be merged.

**Comment by vadimkantorov** [comment]
The current idea is to still have schedulers API specific to learning rate scheduling? (instead of a more generic lr / weight decay schedulers)

**Comment by vincentqb** [comment]
> The current idea is to still have schedulers API specific to learning rate scheduling? (instead of a more generic lr / weight decay schedulers)

The change to chainable scheduler is BC breaking, and so is the change of names (for `get_lr`, `get_last_lr`) if we go the route of generic scheduler. This PR only addresses the first issue of chainability, and is not making an opinion on the latter.

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Yeah, I agree with you. However, "repetition" has been the strategy used in other parts of schedulers to keep them readable, see each block in step that has been repeated.

File: test/test_optim.py
Comment (vincentqb) [reply test/test_optim.py]: Thanks for catching this. I'll remove them.

File: test/test_optim.py
Comment (vincentqb) [reply test/test_optim.py]: Interesting idea :) but we can already change dynamically the schedulers simply using python conditional without a new CombineScheduler, no?

Is the CombineScheduler you propose different from what we discussed [before](https://github.com/pytorch/pytorch/pull/21800#issuecomment-510220723)?

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: The reason I went for `get_last_lr` was to also emphasize that `get_lr` does not return what the user would expect from it, the last computed lr. If there's an `lr` property, then it looks like `get_lr` should return the learning rate.

I originally wanted to change the name of `get_lr`, but that led to many BC breaking change internally, since `get_lr` is used to compute the new learning rate in `step`.

I also wanted to change the names to "values" to remove the reference to "lr" so as to be more general later (so that we can have weight schedulers with identical syntax). I'll leave that for a separate PR.

Thoughts?

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: > However, I am _also_ suggesting that you change the line:
> 
> ```
>         for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
> ```
> 
> To be:
> 
> ```
>         for param_group, lr in zip(self.optimizer.param_groups, self.get_lr_nowarn()):
> ```
> 
> where `get_lr_nowarn` is a version of `get_lr` that has the same semantics, but doesn't do a warning.

I see, so you are proposing (1) `step` uses `get_lr_no_warning` which computes learning rate without warning, (2) `get_lr` raises a warning and calls `get_lr_no_warning`.

However, users in the wild already overwrite `get_lr` with their own, and expect `step` to invoke it without modification. What I describe would then break those codes. Thoughts?

> I think another way of doing this would be with a context manager?

Fair :)

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: It could, yes. The reason I added it in is that LambdaLR felt a little out of place alone since it was not chainable. Do you recommend doing so?

File: test/test_optim.py
Comment (vadimkantorov) [reply test/test_optim.py]: I also think more explicit form of chaining is more understandable over creating multiple schedulers and chaining happening implicitly

File: torch/optim/lr_scheduler.py
Comment (ezyang) [reply torch/optim/lr_scheduler.py]: > Fair :) I'll use that.

Context manager is a strict improvement because it will handle exceptions correctly.

Comment (ezyang) [reply torch/optim/lr_scheduler.py]: > However, users already overwrite get_lr with their own, and expect step to invoke it without modification. What I described just now would then break those codes. Thoughts?

Err, if this is true, you're not going to get warnings in this case, because they've overridden your warnings? It is getting a bit hard to keep track of all the cases. Is there a clear description of how these changes interact with user defined schedulers?

Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: > Err, if this is true, you're not going to get warnings in this case, because they've overridden your warnings? It is getting a bit hard to keep track of all the cases. Is there a clear description of how these changes interact with user defined schedulers?

I don't need to provide a warning in this case. My goal is just to alert to the changes that happened.

* Users should be alerted that some existing schedulers changed from closed form to chainable form.
* Users should be encouraged to use `get_last_lr` to get the last computed value instead of `get_lr` (e.g. by the warning discussed here), since the modified `get_lr` wouldn't returned the last lr for chainable schedulers. I can't make assumptions on how this would behave for a user-modified scheduler.
* Users should be discouraged from providing epoch when not needed.
* Users modified schedulers should continue to work as expected.

> Context manager is a strict improvement because it will handle exceptions correctly.

Agree.

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: MultiplicativeLR moved to #27254

**Comment by vincentqb** [comment]
Since the tests have passed (including the one mentioned [above](https://github.com/pytorch/pytorch/pull/26423#issuecomment-536102835)), I'll be landing this EOD today unless I hear otherwise :)

**Comment by vadimkantorov** [comment]
Are the schedulers still naming-specific to « epochs » as opposed to more generic “steps”?

**Comment by vincentqb** [comment]
> Are the schedulers still naming-specific to « epochs » as opposed to more generic “steps”?

I do agree that the word "epoch" could be replaced by a more fitting one. This PR is not making an opinion on this, besides deprecating its use when not needed.

**Comment by vincentqb** [comment]
Adding "When loading the state of a scheduler, the optimizer also needs to be restored." to BC-breaking, as discussed in #29697.

**Comment by franchesoni** [comment]
Any clue on which schedulers can be concatenated with ReduceLROnPlateau?

If I use CosineAnnealingLR instead of CyclicLR then the code works, but there is the following missing functionality. How could I change the max learning rate in CyclicLR or in CosineAnnealingLR with the same criteria as implemented by ReduceLROnPlateau?

```python
import torch
from torch.nn import Parameter
from torch.optim import SGD

model = [Parameter(torch.randn(2, 2, requires_grad=True))]
optimizer = SGD(model, 0.1)

# scheduler1 = ExponentialLR(optimizer, gamma=0.9)
scheduler1 = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=100)
scheduler2 = torch.optim.lr_scheduler.CyclicLR(optimizer, 0.1, 1, step_size_up=100)

values1, values2 = [], []
for epoch in range(1000):
    values1.append(scheduler2.get_last_lr()[0])
    optimizer.step()
    scheduler1.step(epoch)
    scheduler2.step()

import matplotlib.pyplot as plt
plt.plot(values1, label='values 1')
plt.legend()
plt.show()
```

**Pull request closed without merging**
